In [ ]:
import pandas as pd

# 8.1 Parsing Unix timestamps

It's not obvious how to deal with Unix timestamps in pandas -- it took me quite a while to figure this out. The file we're using here is a popularity-contest file I found on my system at `/var/log/popularity-contest`.

Here's an [explanation of how this file works](http://popcon.ubuntu.com/README).

I'm going to hope that nothing in it is sensitive :)

In [ ]:
full_path_data =  "../data/popularity-contest" #local path will override this if using colab

# Comment out Following 4 line if running the repo locally:

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive') 

Mounted at /content/gdrive


In [ ]:
pwd

'/content'

In [ ]:
ls "gdrive/MyDrive/Teaching and Mentoring/Python/2. Python (cont.)/n. Pandas/Notebooks/Pandas-Complete-Tutorials-Interns/Part-2-Extended-Course/data"

311-service-requests.csv  popularity-contest  weather_2012.csv
bikes.csv                 README.md           weather_2012.sqlite


In [ ]:
full_path_data = "gdrive/MyDrive/Teaching and Mentoring/Python/2. Python (cont.)/n. Pandas/Notebooks/Pandas-Complete-Tutorials-Interns/Part-2-Extended-Course/data/popularity-contest"

In [ ]:
# Read it, and remove the last row
popcon = pd.read_csv(full_path_data, sep=' ', )[:-1]
popcon.columns = ['atime', 'ctime', 'package-name', 'mru-program', 'tag']

The colums are the access time, created time, package name, recently used program, and a tag

In [ ]:
popcon[:5]

atime  ...             tag
0  1387295797  ...             NaN
1  1387295796  ...             NaN
2  1387295743  ...             NaN
3  1387295743  ...  <RECENT-CTIME>
4  1387295742  ...             NaN

[5 rows x 5 columns]

The magical part about parsing timestamps in pandas is that numpy datetimes are already stored as Unix timestamps. So all we need to do is tell pandas that these integers are actually datetimes -- it doesn't need to do any conversion at all.

We need to convert these to ints to start:

In [ ]:
popcon['atime'] = popcon['atime'].astype(int)
popcon['ctime'] = popcon['ctime'].astype(int)

Every numpy array and pandas series has a dtype -- this is usually `int64`, `float64`, or `object`. Some of the time types available are `datetime64[s]`, `datetime64[ms]`, and `datetime64[us]`. There are also `timedelta` types, similarly.

We can use the `pd.to_datetime` function to convert our integer timestamps into datetimes. This is a constant-time operation -- we're not actually changing any of the data, just how pandas thinks about it.

In [ ]:
popcon['atime'] = pd.to_datetime(popcon['atime'], unit='s')
popcon['ctime'] = pd.to_datetime(popcon['ctime'], unit='s')

If we look at the dtype now, it's `<M8[ns]`. As far as I can tell `M8` is secret code for `datetime64`.

In [ ]:
popcon['atime'].dtype

dtype('<M8[ns]')

So now we can look at our `atime` and `ctime` as dates!

In [ ]:
popcon[:5]

atime  ...             tag
0 2013-12-17 15:56:37  ...             NaN
1 2013-12-17 15:56:36  ...             NaN
2 2013-12-17 15:55:43  ...             NaN
3 2013-12-17 15:55:43  ...  <RECENT-CTIME>
4 2013-12-17 15:55:42  ...             NaN

[5 rows x 5 columns]

Now suppose we want to look at all packages that aren't libraries.

First, I want to get rid of everything with timestamp 0. Notice how we can just use a string in this comparison, even though it's actually a timestamp on the inside? That is because pandas is amazing.

In [ ]:
popcon = popcon[popcon['atime'] > '1970-01-01']

Now we can use pandas' magical string abilities to just look at rows where the package name doesn't contain 'lib'.

In [ ]:
nonlibraries = popcon[~popcon['package-name'].str.contains('lib')]

In [ ]:
nonlibraries.sort_values('ctime', ascending=False)[:10]

atime  ...             tag
57  2013-12-17 04:55:39  ...  <RECENT-CTIME>
450 2013-12-16 20:03:20  ...  <RECENT-CTIME>
454 2013-12-16 20:03:20  ...  <RECENT-CTIME>
445 2013-12-16 20:03:20  ...  <RECENT-CTIME>
396 2013-12-16 20:08:27  ...  <RECENT-CTIME>
449 2013-12-16 20:03:20  ...  <RECENT-CTIME>
397 2013-12-16 20:08:25  ...  <RECENT-CTIME>
398 2013-12-16 20:08:23  ...  <RECENT-CTIME>
452 2013-12-16 20:03:20  ...  <RECENT-CTIME>
440 2013-12-16 20:03:20  ...  <RECENT-CTIME>

[10 rows x 5 columns]

Okay, cool, it says that I I installed ddd recently. And postgresql! I remember installing those things. Neat.

The whole message here is that if you have a timestamp in seconds or milliseconds or nanoseconds, then you can just "cast" it to a `'datetime64[the-right-thing]'` and pandas/numpy will take care of the rest.

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  